gpytorch/variational/_variational_strategy.py

```diff
@@ -0,0 +1,138 @@
+        :param torch.Tensor inducing_points: Locations :math:`\mathbf Z` of the inducing points
+        :param torch.Tensor inducing_values: Samples of the inducing function values :math:`\mathbf u`
+            (or the mean of the distribution :math:`q(\mathbf u)` if q is a Gaussian.
+        :param ~gpytorch.lazy.LazyTensor variational_inducing_covar: If the distribuiton :math:`q(\mathbf u)`
+            is Gaussian, then this variable is the covariance matrix of that Gaussian. Otherwise, it will be
+            :attr:`None`.
+
+        :rtype: :obj:`~gpytorch.distributions.MultivariateNormal`
+        :return: The distribution :math:`q( \mathbf f(\mathbf X))`
+        """
+        raise NotImplementedError
+
+    def kl_divergence(self):
+        """
+        Compute the KL divergence between the variational inducing distribution :math:`q(\mathbf u)`
+        and the prior inducing distribution :math:`p(\mathbf u)`.
+
+        :rtype: torch.Tensor
+        """
+        with settings.max_preconditioner_size(0):
+            kl_divergence = torch.distributions.kl.kl_divergence(
+                self.variational_distribution,
+                self.prior_distribution
+            )
+        return kl_divergence
+
+    def train(self, mode=True):
+        # Make sure we are clearing the cache if we change modes
+        if (self.training and not mode) or mode:
+            if hasattr(self, "_memoize_cache"):
+                delattr(self, "_memoize_cache")
+        return super().train(mode=mode)
+
+    def __call__(self, x, prior=False):
+        # If we're in prior mode, then we're done!
+        if prior:
+            return self.model.forward(x)
+
+        # Delete previously cached items from the training distribution
+        if self.training:
+            if hasattr(self, "_memoize_cache"):
+                delattr(self, "_memoize_cache")
+                self._memoize_cache = dict()
+        # (Maybe) initialize variational distribution
+        if not self.variational_params_initialized.item():
+            prior_dist = self.prior_distribution
+            self._variational_distribution.initialize_variational_distribution(prior_dist)
+            self.variational_params_initialized.fill_(1)
+
+        # Ensure inducing_points and x are the same size
+        inducing_points = self.inducing_points
+        if inducing_points.shape[:-2] != x.shape[:-2]:
+            batch_shape = _mul_broadcast_shape(inducing_points.shape[:-2], x.shape[:-2])
+            inducing_points = inducing_points.expand(*batch_shape, *inducing_points.shape[-2:])
+            x = x.expand(*batch_shape, *x.shape[-2:])
+
+        # Get p(u)/q(u)
+        variational_dist_u = self.variational_distribution
+
+        # Get q(f)
+        if isinstance(variational_dist_u, MultivariateNormal):
+            return super().__call__(
+                x, inducing_points,
+                inducing_values=variational_dist_u.mean,
```

Comment:
Not wild about passing the variational mean through “inducing_values”. Think I know why you did it (`Delta`), but could we not just pass `variational_dist_u` along and have the strategy worry about assumptions about what `variational_dist_u` has in it?

Reply:
That's a fair point. The reason I did it this way is to give us some future flexibility if we wanted to use non-Gaussian variational distributions (e.g. normalizing flows) that we'd have to sample from. Then `inducing_values` would correspond to samples from `q(u)`.

Also, this way the casing logic can happen in the super class.

Reply:
Would any of the existing strategies be usable with samples from a NF? If not, then it still might be better to pass the dist along since a NFVariationalStrategy or whatever will need yet another case added here

Reply:
Atm all of the existing strategies could be used with samples from a NF. Computing q(f | u) (for a sample from u) is the same as computing E[ q(f) ] for delta var dists. is the same as computing E[ q(f) ] for normal var dists.

I agree that it's not the nicest inteface in terms of method names, but it does reduce the amount of repeated code. It's written imagining that the "default" case is getting samples from q(u) (which trivially covers Delta distributions). Normal distributions are written as a special case (where we do marginalization with the covariance), but the code for computing the mean is the same as the "default" case.

Open to suggestions for improving this.

Reply:
I'm kind of just armchair reviewing from my phone, so I'm fine letting it go in as is and opening another issue if it turns out to be problematic down the road.